clc/modules/msgs/src/main/java/com/eucalyptus/ws/Handlers.java

```diff
@@ -175,7 +176,47 @@ private static NioServerHandler newNioServerHandler( ) {
   }
 
   private static ChannelHandler newChunkedWriteHandler( ) {
-    return new ChunkedWriteHandler( );
+    return new ChunkedWriteHandler( ) {
+      private void flush( final ChannelHandlerContext ctx ) {
+        try {
+          beforeRemove( ctx );
+        } catch ( final Exception e ) {
+          LOG.warn( "Unexpected exception while sending chunks.", e );
+        }
+      }
+
+      @Override
+      public void handleUpstream( final ChannelHandlerContext ctx, final ChannelEvent e ) throws Exception {
+        if ( e instanceof ChannelStateEvent ) {
+          final ChannelStateEvent cse = (ChannelStateEvent) e;
+          if ( INTEREST_OPS == cse.getState( ) ) {
+            // flush off the io thread to avoid blocking other channels
+            pipelineExecutionHandler( ).getExecutor( ).execute( () -> flush( ctx ) );
+            ctx.sendUpstream(e);
+            return;
+          }
+        }
+        super.handleUpstream( ctx, e );
+      }
+
+      @Override
+      public void handleDownstream( final ChannelHandlerContext ctx, final ChannelEvent e ) throws Exception {
+        final Channel channel = ctx.getChannel();
+        final boolean moveFlush = e instanceof MessageEvent && channel.isWritable( );
+        if ( moveFlush ) {
+          final Channel flushSuppressingChannel = new DelegatingChannel( channel ) {
+            @Override public boolean isWritable( ) { return false; }
+          };
+          super.handleDownstream( new DelegatingChannelHandlerContext( ctx ) {
+            @Override public Channel getChannel( ) { return flushSuppressingChannel; }
+          }, e );
+          // flush off the io thread to avoid blocking other channels
+          pipelineExecutionHandler( ).getExecutor( ).execute( () -> flush( ctx ) );
+        } else {
+          super.handleDownstream( ctx, e );
+        }
+      }
```

Comment:
Here is the super impl:

```
    public void handleDownstream(ChannelHandlerContext ctx, ChannelEvent e)
            throws Exception {
        if (!(e instanceof MessageEvent)) {
            ctx.sendDownstream(e);
            return;
        }

        boolean offered = queue.offer((MessageEvent) e);
        assert offered;

        final Channel channel = ctx.getChannel();
        // call flush if the channel is writable or not connected. flush(..) will take care of the rest

        if (channel.isWritable() || !channel.isConnected()) {
            this.ctx = ctx;
            flush(ctx, false);
        }
    }
```

So if we need to move the flush off of the io thread the we suppress flush on the channel when calling super#handleDownstream and then call flush. If moving flush is not needed then we directly call the super implementation.